packages/serve/src/lib/catalog-router/catalogRouters.ts

```diff
@@ -0,0 +1,103 @@
+import { Next } from 'koa';
+import * as Router from 'koa-router';
+import { CatalogRouter } from '@vulcan-sql/serve/models';
+import { inject } from 'inversify';
+@VulcanInternalExtension('catalog')
+@VulcanExtensionId('catalog')
+export class CatalogRouters extends CatalogRouter {
+  private router = new Router();
+
+  constructor(
+    @inject(CORE_TYPES.ExtensionConfig) config: any,
+    @inject(CORE_TYPES.ExtensionName) moduleName: string,
+    @inject(CORE_TYPES.ProjectOptions) projectOptions: Partial<ICoreOptions>,
+    @inject(CORE_TYPES.ArtifactBuilder) artifactBuilder: VulcanArtifactBuilder
+  ) {
+    super(config, moduleName, projectOptions, artifactBuilder);
+  }
+
+  public override async onActivate() {
+    this.router.get('/catalog/schemas/:sourceName', async (ctx, next) => {
+      await next();
+      const { sourceName } = ctx.params;
+      const schemas = await this.getArtifactSchemas();
+      const [schema] = schemas.filter((schema) => {
+        return (
+          (schema as APISchema & { sourceName: string }).sourceName ===
+          sourceName
+        );
+      });
+      if (!schema) {
+        ctx.response.status = 404;
+        ctx.response.body = 'Not Found';
+        return;
+      }
+
+      const baseUrl = `${ctx.protocol}://${ctx.host}`;
+      const result = {
+        ...schema,
+        url: `${baseUrl}/api${schema.urlPath}`,
+        apiDocUrl: `${baseUrl}${this.getAPIDocUrl(schema)}`,
+        shareKey: this.getShareKey(ctx.request.headers.authorization),
+      };
+      ctx.response.body = result;
+    });
+
+    this.router.get('/catalog/schemas', async (ctx, next) => {
+      await next();
+      const schemas = await this.getArtifactSchemas();
+      const baseUrl = `${ctx.protocol}://${ctx.host}`;
+      const result = schemas.map((schema) => {
+        return {
+          ...schema,
+          url: `${baseUrl}/api${schema.urlPath}`,
+          apiDocUrl: `${baseUrl}${this.getAPIDocUrl(schema)}`,
+          shareKey: this.getShareKey(ctx.request.headers.authorization),
+        };
+      });
+      ctx.response.body = result;
+    });
+  }
+
+  private getShareKey(authorization: string | undefined) {
+    if (!authorization) return '';
+
+    const project = this.getProjectOptions();
+    const authSource = project?.['auth-source'];
+    const token = Buffer.from(
+      JSON.stringify({ Authorization: authorization })
+    ).toString('base64');
+    const key = (authSource && authSource?.options?.key) || 'auth';
+    return `?${key}=${token}`;
+  }
+
+  private getAPIDocUrl(schema: APISchema) {
+    const splitUrl = schema.urlPath.split('/');
+    const restructureUrl = splitUrl
+      .map((brick) =>
+        brick.startsWith(':') ? `{${brick.replace(':', '')}}` : brick
+      )
+      .join('~1');
+    const project = this.getProjectOptions();
+    const redoc = project?.['redoc'];
+    const docPath =
+      redoc?.url?.replace(/\/+$/, '').replace(/^\/+/, '') || 'doc';
```

Comment:
Why should need to add `+` in the first `replace` and send `replace` method? seems you would like to remove the first and last `/` ? 

Btw, the same part also appeared in the `AuthCredentialsMiddleware` and `AuthSourceNormalizerMiddleware`, but you not add `+`

Reply:
Thanks for finding the difference, I just copied the original code from document router,
It will move to a function and provide the auth refactor function you mention below to use.